### PR TITLE
Implement stream compression lib module

### PR DESCRIPTION
**Pull request by muxi**
The first portion of stream compression efforts. Creates the compression/decompression lib module to be used by gRPC transport. The API may act as polymorphism interface in the future when more algorithms are supported.

### CODE REVIEW

File: test/core/compression/stream_compression_test.c [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+  grpc_slice slice = grpc_slice_from_static_string(test_str);
+  grpc_slice_buffer_add(&source, slice);
+  GPR_ASSERT(grpc_stream_compress(compress_ctx, &source, &relay, NULL,
+                                  ~(size_t)0,
+                                  GRPC_STREAM_COMPRESSION_FLUSH_FINISH));
+  bool end_of_context;
+  size_t output_size;
+  GPR_ASSERT(grpc_stream_decompress(decompress_ctx, &relay, &sink, &output_size,
+                                    ~(size_t)0, &end_of_context));
+  GPR_ASSERT(output_size = sizeof(test_str) - 1);
```
Comment (dgquintas) [inline test/core/compression/stream_compression_test.c]: is this supposed to be a ==? You are assigning, so this assert will pass as long as test_str != 1

File: test/core/compression/stream_compression_test.c [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+  grpc_slice slice_recv = grpc_slice_buffer_take_first(&sink);
+  char *str_recv = (char *)GRPC_SLICE_START_PTR(slice_recv);
+  GPR_ASSERT(GRPC_SLICE_LENGTH(slice_recv) == max_output_size);
+  GPR_ASSERT(0 == strncmp(test_str, str_recv, max_output_size));
+  grpc_slice_unref(slice_recv);
+
+  size_t remaining_size = sizeof(test_str) - 1 - max_output_size;
+  GPR_ASSERT(grpc_stream_decompress(decompress_ctx, &relay, &sink, &output_size,
+                                    remaining_size, &end_of_context));
+  GPR_ASSERT(output_size = remaining_size);
```
Comment (dgquintas) [inline test/core/compression/stream_compression_test.c]: you keep using = when you need ==, all throughout this test.

File: src/core/lib/compression/stream_compression.h [hunk 3]
```diff
@@ -31,14 +31,15 @@ typedef struct grpc_stream_compression_context {
 } grpc_stream_compression_context;
 
 typedef enum grpc_stream_compression_method {
-  GRPC_STREAM_COMPRESSION_COMPRESS,
+  GRPC_STREAM_COMPRESSION_COMPRESS = 0,
   GRPC_STREAM_COMPRESSION_DECOMPRESS
```
Comment (dgquintas) [inline src/core/lib/compression/stream_compression.h]: no COUNT?

**Comment by muxi** [comment]
test this please

**Comment by muxi** [comment]
test this please

**Comment by muxi** [comment]
Test failure:
Asan C++ - #10530 
Basic Tests Mac - #9562
gRPC_interop_pull_requests - #11857

**Pull request merged**